Optional Timer Revisited: Works toward #40

andrewhansbury opened the pull request:
For anyone wondering, this pull request is a continuation of https://github.com/Deadpikle/pathfinder-bible-experience-prep-quiz-master/pull/64
I am unfortuanetly a git-newbie, so I'm not sure if opening this secondary pull request is the proper way to go about fixing the issues in my first. My gut tells me it probably should have been done in the single request. Please let me know if this needs to be merged into https://github.com/Deadpikle/pathfinder-bible-experience-prep-quiz-master/pull/64.

# Optional Timer.... Improved!

I am going to structure this pull request to match the structure of the commentary given in https://github.com/Deadpikle/pathfinder-bible-experience-prep-quiz-master/pull/64.

## Non-Code
Thanks to the incredibly thorough pull request review by @Deadpikle, say hello to the new and improved Optional-Timer! I am learning a whole lot on how git works in a team setting, which is an invaluable skill that I really needed to start learning, so I am super thankful for this opportunity! "A lot of text below" just meant a lot of stuff to learn!

### Some standard practice
This pull request is now referencing the original issue, https://github.com/Deadpikle/pathfinder-bible-experience-prep-quiz-master/issues/40. I have it titled as "Works Toward" https://github.com/Deadpikle/pathfinder-bible-experience-prep-quiz-master/issues/40, but hopefully we are getting closer to "Closes https://github.com/Deadpikle/pathfinder-bible-experience-prep-quiz-master/issues/40."
Is there any other GitHub conventions I am messing up?

### Git
I tried to break up these commits into much smaller pieces than last time. I have 10 commits for this PR!

## Code
### Files for the library
Adding just the necessary file certainly cleaned up this PR a whole lot! I didn't realize the only necessary file was `easytimer.min.js`, even though it specifies the direct path to it 😄.

I found the other script references in the code base and followed their example. The added easytimer.twig file points to the timer library in the `lib/` directory.

### Common.css
I'm glad I dont have to Google how to disable linter in VS Code 😂.

### Overall Code
Once again, for these changes I tried to maintain the style of the rest of the codebase to the best of my ability.

## Improvement requests / Moving Forward
Great ideas for the improvements! I think I was able to get most of them implemented 👍🏻

- Question timer is now hidden during initial countdown
- Initial countdown text is hidden when the question timer starts
- Icon comes in with the question timer. I added the materializecss link you provided in the same file as the other icons
- I wasn't able to replicate the -1 bug. Not sure about how or why that happened.
- "Show Answer" button now hides the timer stuff!
-  Why not both! I went ahead and added the opt-out option for auto-showing the answer when the timer finishes. 
- I certainly see your concern with the wall of text. I do think the timer icon we just added makes it look much better now, and in addition, there is one less line of text since the initial 10-second coundown is replaced with the question timer once the timer begins. Let me know if you still think more needs to be done. Maybe we could combine the points possible line with the timer line, but that might be unintuitive.  We could also potentially position the timer elsewhere, like in the center of screen. 

### Potential bugs.... Squashed! (I hope)
- Good catch on my potential array out of bounds. I think it should be covered now. 0 point questions are given 25 seconds
- `if (data.points > 8){
                    var questionLength = 60 + ((data.points - 8) * 15);
                }` Some fun math to handle questions worth more than 8 points.
___
I think that's everything! Looking forward to code review part 2!!

Deadpikle commented:
:wave: Sorry for the wait, here. If my slowness stops your educational needs (I am swamped with way too much right now), please please please ping me one way or another and I will get on things faster.

## Non code things

> New PR vs old PR

So...a couple ways this could have gone. Opening a new PR when there are major substantive changes (such as a bunch of new commits) is not necessarily wrong, but especially on larger repositories with more work being done, this could be considered "spam" potentially (opening and closing PRs for the same task). So, usually you will want to keep things on the same PR.

To update the code for a PR, you can just push to whatever branch you originally created the PR from. In your original case, that would mean pushing commits to the `optional-timer` branch on your repository.

"But wait, I've redone a bunch of commits and can't just push!"

Well...yeah...`git push -f`(force push) is a thing to be used sparingly but wisely. When redoing major things for a PR, you can force-push to override the code in the PR if there are huge changes. Just make sure your work is backed up locally before a force push juuuuust in case.

The other thing that some repo maintainers might do is squash all your commits into one commit when the PR is accepted. If they do that, then it won't matter what order your commits are in -- it'll get merged in as 1 commit. But doing anything you can to make reviewing your changes easier for others is a good thing. Just don't erase your work at any time until you know it's all safe 🤪

Moving forward from here, just keep things in this PR.

> I tried to break up these commits into much smaller pieces than last time. I have 10 commits for this PR!

👍🏻 

>  Is there any other GitHub conventions I am messing up?

Replacing "messing up" with "learning" 😉 

Not anything in particular for this repo since it's fairly small. Some repositories will have templates for Issues or Pull Requests, which gives users a template to fill out when filing a request so that all the pertinent information is there for a PR or issue. See [this example of an issue template](https://github.com/AvaloniaUI/Avalonia/blob/master/.github/ISSUE_TEMPLATE/bug_report.md) and [this example of a pull request template](https://github.com/AvaloniaUI/Avalonia/blob/master/.github/PULL_REQUEST_TEMPLATE.md) for examples. Not using a template when one is available will make people Not Happy, but there are no templates on this repo, so you're good.

On other repos on projects that have this set up, PRs will trigger unit tests and other things to run. It's good practice that if your code breaks some unit tests, then you should fix the unit tests so that everything passes! (Also, those repos will probably want unit tests to be written for your code. Again, this repo isn't doing all that right now, so you're fine.)

When doing anything that touches the GUI, screenshots and/or GIFs/videos are usually very helpful. Don't worry about that for now, but usually they help people know what's going on without building/running the code themselves.

Your changelog of what happened between your PRs (and in the normal case, between major changes in the same PR) is a good thing. Well done on that.

For an example of what this might look like on a big repo, see [this PR](https://github.com/AvaloniaUI/Avalonia/pull/6024) or [this PR](https://github.com/AvaloniaUI/Avalonia/pull/6024) (both by me just because 🤷🏻‍♂️)

## Change requests/Bugs

I think the "wall of text" is greatly helped with the icon, like you said. 👍🏻

* [ ] Swap the order of the timer icon and the countdown timer so that it's [icon] [number]. You might need to put the `<i>` inside the `<h5>`? (That'd probably be better anyway.)
* [ ] `#timer-icon` CSS -- add `vertical-align: text-bottom;` so that things line up nicely
* [ ] Let's tweak the countdown text for the 10 second timer to `25 second timer starts in # seconds...` (and make sure it changes to "second" with 1 second
* [ ] Only make the timer text red once it hits 10 seconds (black color the rest of the time)
* [ ] The two options on the setup quiz screen can be in the same "section" labeled "Question timer" (in other words, you can get rid of the "Auto-show answer" header)
* [ ] Bug: Disable question timer. First question - OK. Second question: A mysterious "10" label shows up.

<img width="537" alt="Screen Shot 2022-05-04 at 10 00 21 PM" src="https://user-images.githubusercontent.com/5092399/166853852-54fd403a-bf36-4f7f-9aad-781b82d3921e.png">

* [ ] The `{% include 'includes/scripts/easytimer.twig' %}` line needs to be in the `headers` block, not the main content block. This way it'll end up in the `<head>` of the document rather than the `<body>`. Browsers don't complain, but it's good practice. E.g.,

```twig
{% block headers %}

{% include 'includes/scripts/easytimer.twig' %}

{% endblock %}
```

Deadpikle commented on views/user/quiz/take-quiz.twig:
```diff
@@ -186,6 +197,40 @@
+                    var showAnswer = document.getElementById('show-answer');
+                    showAnswer.click();
+                }
+                else{
```
Nit: space between `else` and `{`

Deadpikle commented on views/user/quiz/take-quiz.twig:
```diff
@@ -186,6 +197,40 @@
+            });
+            // Simulating "Show Answer" button click when timer finishes
+            timer.addEventListener('targetAchieved', function (e) {
+                if (!disableAutoShowAnswer){
```
Nit: space between `)` and `{`

Deadpikle commented on views/user/quiz/take-quiz.twig:
```diff
@@ -186,6 +197,40 @@
+                $("#timer-icon").html("timer");
+                timer.stop();
+                timer.start({countdown:true, precision: 'seconds', startValues: {seconds: timerLength}}); 
+     
```
Nit: unneeded whitespace

Deadpikle commented on views/user/quiz/take-quiz.twig:
```diff
@@ -335,12 +381,22 @@
             }
         }, false);
+        
+        function stopTimer(){
```
Nit: space between `)` and `{` (and elsewhere)

andrewhansbury commented:
## Updated Commits

Hi @Deadpikle,
I believe I have resolved all your comments on this pull request!
Looking forward to getting some more feedback! :)

Deadpikle commented:
@andrewhansbury -- Looks fantastic. One very minor tweak and one minor UI fix -> fix those and I'll merge this PR in and push the update live for all users!

* [ ] `#question-timer` CSS -> `vertical-align: text-bottom` (like the icon -- it centers things nicely)
* [ ] There's an odd gap that shouldn't be here when the timer isn't showing or when it's finished. Especially noticeable when not showing the timer at all.

<img width="279" alt="Screen Shot 2022-05-18 at 7 38 13 PM" src="https://user-images.githubusercontent.com/5092399/169172403-9ed2b41c-f619-4427-bcc8-a8dea7a93b0d.png">

<img width="552" alt="Screen Shot 2022-05-18 at 7 38 44 PM" src="https://user-images.githubusercontent.com/5092399/169172441-b820498b-fb44-48f6-a22d-d1f666da262d.png">

andrewhansbury commented:
Hey @Deadpikle,

Glad you liked the changes,
these last commits should have fixed the tweaks you mentioned!

P.S.,  
I added one last surprise feature in that last commit, I hope you like it!
The file used was found [here](https://www.dreamstime.com/royalty-free-sound-effect-short-ding-bell-sound-use-as-new-message-video-games-etc-fairy-ding-audio118782333), and is royalty-free, so it should be good for use on this project. It could also be easily swapped with any other sound you would like, just change the file I added to the /files/ folder.

Hopefully I didn't break anything when making these updates!

Thanks,
- @andrewhansbury

Deadpikle commented:
Thank you! 

> P.S., I added one last surprise feature in that last commit, I hope you like it!

I don't really want this change, so I have reverted it.

----

Your changes are approved, and I am merging this in and publishing the feature now. Thank you very much!!!!
